overlord/snapstate/handlers.go

```diff
@@ -1771,8 +1773,22 @@ func (m *SnapManager) stopSnapServices(t *state.Task, _ *tomb.Tomb) error {
 	st.Lock()
 	defer st.Unlock()
 
-	// finally commit the disabled services to snapsetup
-	snapsup.LastActiveDisabledServices = disabledServices
+	// for undo
+	t.Set("old-last-active-disabled-services", snapst.LastActiveDisabledServices)
+	// undo could queryDisabledServices, but this avoids it
+	t.Set("disabled-services", disabledServices)
+
+	// add to the disabled services list in snapst services which were disabled
+	// for usage across changes like in reverting and enabling after being
+	// disabled.
+	// we keep what's already in the list in snapst because that list is
+	// services which were previously present in the snap and disabled, but are
+	// no longer present.
+	snapst.LastActiveDisabledServices = append(
+		snapst.LastActiveDisabledServices,
+		disabledServices...,
+	)
+	Set(st, snapsup.InstanceName(), snapst)
 
 	err = SetTaskSnapSetup(t, snapsup)
```

Comment:
this bit is not needed anymore, we are not writing into snapsup atm

Reply:
Indeed, thanks, removed.